flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java

```diff
@@ -249,7 +249,7 @@ public Builder uidPrefix(String newPrefix) {
       return this;
     }
 
-    public DataStreamSink<RowData> build() {
+    public DataStreamSink<Void> build() {
```

Comment:
technically, this is a breaking change if users store the return value into a variable, although in theory I can't think of a valid reason to do so.
```
    DataStreamSink<RowData> sink = FlinkSink.forRowData(dataStream)
        ...
        .build();
    // do sth with the sink like override/set name, uid etc.
```

Reply:
Since this isn't going to return anything anyway, why not let the user see whatever type they want?

```java
  public <T> DataStreamSink<T> build() {
    ...
    return (DataStreamSink<T>) appendDummySink(...);
  }
```

Reply:
On similar track as @rdblue 's suggestion, we can mark the current `build()` API as deprecated so that we don't break the backward compatibility. Then we can add a new `append()` API 
```
public void append() {
  appendDummySink(...);
}
```

Reply:
@stevenzwu , Yes, it's a breaking changes for flink DataStream API, it's necessary to maintain the compatibility for at least one major iceberg release.    @rdblue ,  for flink DataStream operators chain,  the returned DataStreamSink should has the same return type as the last chained operator,  so it does not make sense to return a DataStreamSink with an arbitrary type as user want.

For @stevenzwu 's second suggestion,   I think it's still necessary to return a `DataStreamSink<Void>` in case of people want to set their own `name`, `uid`, `ChainingStrategy` etc, though less people will really do that but we need to.

Reply:
>  it's still necessary to return a DataStreamSink<Void> in case of people want to set their own name, uid, ChainingStrategy

I am not sure if we want to enable users to do that. The dummy sink is part of the Flink Iceberg sink DAG: map (optional) -> writer -> committer -> dummy sink. We don't allow users to change the properties (like name, uid, chaining strategy, parallelism, slot sharing group) for other operators. why do we want to allow the change only for the dummy sink operator.

But I am not against returning `DataStreamSink<Void>`, as in theory it is natural. I just have some doubt on the practical values of exposing it.

Reply:
Hi @stevenzwu ,   we need to return a `DataStreamSink<Void>` because in the [IcebergTableSink](https://github.com/apache/iceberg/blob/a0d6cdf8a76e29d07ce0d5889b7ee9a2335bde1f/flink/src/main/java/org/apache/iceberg/flink/IcebergTableSink.java#L62), we will need to return a `DataStreamSink` to implement the [DataStreamSinkProvider](https://github.com/apache/flink/blob/db5af67aee7fb449a9db8930213eec7dc925e58c/flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/connector/sink/DataStreamSinkProvider.java).